### ISSUE TITLE
ImportError: cannot import name 'TypeGuard' from 'typing_extensions'

### ISSUE THREAD

**Issue by IndigoWizard** [python/typing_extensions#89]
This may be more suited for [Streamlit](https://github.com/streamlit/streamlit) (I did open one at [streamlit/streamlit#5551](https://github.com/streamlit/streamlit/issues/5551) ) so if it shouldn't be opened here feel free to close/delete this issue.

### Summary
When I try to run my [streamlit](https://github.com/streamlit/streamlit) project app by launching the local server `streamlit run app.py` it doesn't work and instead sends back the error: **` ImportError: cannot import name 'TypeGuard' from 'typing_extensions' `**

### Session log:

```python
(geospatials) PS C:\Users\DELL\Documents\GitHub\My-Forks\streamlit-project> streamlit run app.py
Traceback (most recent call last):
  File "C:\ProgramData\Anaconda3\envs\geospatials\Scripts\streamlit-script.py", line 5, in <module>
    from streamlit.web.cli import main
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\__init__.py", line 70, in <module>
    from streamlit.delta_generator import DeltaGenerator as _DeltaGenerator
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\delta_generator.py", line 36, in <module>
    from streamlit import cursor
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\cursor.py", line 18, in <module>
    from streamlit.runtime.scriptrunner import get_script_run_ctx
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\runtime\scriptrunner\__init__.py", line 16, in <module>
    from .script_runner import (
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\runtime\scriptrunner\script_runner.py", line 35, in <module>
    from streamlit.runtime.state import (
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\runtime\state\__init__.py", line 16, in <module>
    from .safe_session_state import SafeSessionState as SafeSessionState
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\runtime\state\safe_session_state.py", line 21, in <module>
    from .session_state import (
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\runtime\state\session_state.py", line 44, in <module>
    from streamlit.type_util import ValueFieldName
  File "C:\ProgramData\Anaconda3\envs\geospatials\lib\site-packages\streamlit\type_util.py", line 35, in <module>
    from typing_extensions import (
ImportError: cannot import name 'TypeGuard' from 'typing_extensions' (C:\Users\DELL\AppData\Roaming\Python\Python39\site-packages\typing_extensions.py)
```

Also it seems that two packages with similar names are installed as part of Streamlit dependencies when using Conda; Are both "typing_extensions" and "typing-extensions" the same package or different ones?

![image](https://user-images.githubusercontent.com/43890965/196054408-805f56f9-32e5-4a08-98ca-39c2ebbc4dc9.png)

### Debug info

- Streamlit version: 1.13.0
- Python version: 3.9.13
- Using Conda: 22.9.0
- OS version: Windows 10
- Browser version: Firefox 105.0.3 (64-bit)

### Additional info:
I already tried creating new conda envs with different python versions (`3.8.3` & `3.9.13`) in addition to the latest one in the screenshot (`3.9.12`).
I also tried installing both latest streamlit (`1.13.0`) and default one in conda (`1.11.0`) I always get the same error and both "**typing_extensions**" and "**typing-extensions**" are in version 4.4.0 in all the cases.
Tried all following versions of "**typing_extensions**" and "**typing-extensions**" and still had same error:
- `4.3.0`
- `4.2.0`
- `4.1.1`
- `3.10.0.2`

**Comment by gvanrossum** [comment]
Let's see what the Streamlit folks say. You can reopen if they say it's not their fault. (Or add a comment to reopen if GitHub won't let you -- I'll keep monitoring this issue, no need to create a new one.)

**Comment by gvanrossum** [comment]
PS. It's also possible that this is due to conda? They have a tracker too. :-)

**Comment by JelleZijlstra** [comment]
FYI, `TypeGuard` has been in typing-extensions since before 4.0, which is when we started keeping a changelog. The underscore/dash confusion in the name may be because of a packaging change a while ago that affected the project's name in some packaging artifacts.

**Comment by IndigoWizard** [comment]
As you can see in the links in front of each package & version, I did find that there are both packages on both `anaconda` & `conda-forge` trackers and presented with different versions :

#### anaconda
- [`typing-extensions 4.3.0`](https://anaconda.org/anaconda/typing-extensions)
- [`typing_extensions 4.3.0`](https://anaconda.org/anaconda/typing_extensions)

#### conda-forge 
- [`typing-extensions 4.4.0`](https://anaconda.org/conda-forge/typing-extensions)
- [`typing_extensions 4.4.0`](https://anaconda.org/conda-forge/typing_extensions)

All these packages on `anaconda` and `conda-forge` trackers do link to this very same project [python/typing_extensions](https://github.com/python/typing_extensions/).

Wouldn't it be wise to delete one of the packages on each tracker and keep only the one with the correct name and the latest version like it is the case on `pypi` tracker [`typing-extensions 4.4.0`](https://pypi.org/project/typing-extensions/) where there is only one package?
I'm asking here the maintainers of the project as I believe it's more suitable for you guys to signal the issue on `anaconda` & `conda-forge` trackers to have things setup right.

**Comment by BastianZim** [comment]
Conda-forge maintains both versions as some users misspell the name and confuse `-` and `_`. This makes it easier to install and in line with pip which doesn't distinguish between `-` and `_`.

If you install `typing-extensions` from conda-forge, you get an empty package called `typing-extensions` and additionally `typing_extensions` is installed as well, giving you the functionality. This is deliberate. 

In short, this is only a placeholder and does not affect anything.

The error should likely be from something else and not the different names.

Ref: https://github.com/conda-forge/typing_extensions-feedstock/blob/main/recipe/meta.yaml
